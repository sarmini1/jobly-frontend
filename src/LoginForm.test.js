import React from 'react';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from 'react-dom/test-utils';
import { MemoryRouter } from "react-router-dom";
import LoginForm from './LoginForm';

import { fireEvent } from '@testing-library/react';


let container = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  // console.log(document);
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('mounts without crashing', async function () {
  await act(async () => {
    render(
      <MemoryRouter>
        <LoginForm logIn={() => { }} />
      </MemoryRouter>
      , container);
  });
});

it("matches snapshot", async function () {
  await act(async () => {
    render(
      <MemoryRouter>
        <LoginForm logIn={() => { }} />
      </MemoryRouter>
      , container);
  });
  expect(container).toMatchSnapshot();
});

it("renders empty form successfully", async function () {

  // Use the asynchronous version of act to apply resolved promises
  await act(async () => {
    render(
      <MemoryRouter>
        <LoginForm logIn={() => { }} />
      </MemoryRouter>
      , container);
  });

  const usernameInput = container.querySelector("#LoginForm-username");
  const passwordInput = container.querySelector("#LoginForm-password");

  expect(container).toContainHTML("<h3>Log in</h3>");
  expect(container).toContainElement(usernameInput);
  expect(container).toContainElement(passwordInput);

});

it("calls prop function upon submission", async function () {
  const mockLogIn = jest.fn();

  await act(async () => {
    render(
      <MemoryRouter>
        <LoginForm logIn={mockLogIn} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#LoginForm-username");
  const passwordInput = container.querySelector("#LoginForm-password");

  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "test-username" } });
    fireEvent.change(passwordInput, { target: { value: "test-password" } });
    fireEvent.submit(form);
  });

  expect(mockLogIn).toHaveBeenCalledWith(
    { username: "test-username", password: "test-password" }
  );
  expect(mockLogIn).toBeCalledTimes(1);

  mockLogIn.mockRestore();
});

it("shows errors when submission function throws them", async function () {
  const mockLogIn = jest
    .fn()
    .mockImplementationOnce(() => Promise.reject(["mock error message"]));

  await act(async () => {
    render(
      <MemoryRouter>
        <LoginForm logIn={mockLogIn} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#LoginForm-username");
  const passwordInput = container.querySelector("#LoginForm-password");

  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "test-username" } });
    fireEvent.change(passwordInput, { target: { value: "test-password" } });
    fireEvent.submit(form);
  });

  expect(mockLogIn).toHaveBeenCalledWith(
    { username: "test-username", password: "test-password" }
  );
  expect(mockLogIn).toBeCalledTimes(1);

  // shows errors
  const error = container.querySelector(".Error");
  expect(error).toContainHTML("mock error message");
  expect(container).toContainElement(error);

  // holds onto initial inputs in form
  expect(usernameInput).toContainHTML("test-username");

  mockLogIn.mockRestore();
});


it("can resubmit form when errors are present", async function () {
  const mockLogIn = jest
    .fn()
    .mockImplementationOnce(() => Promise.reject(["mock error message"]));

  await act(async () => {
    render(
      <MemoryRouter>
        <LoginForm logIn={mockLogIn} />
      </MemoryRouter>
      , container);
  });

  const form = container.querySelector("form");
  const usernameInput = container.querySelector("#LoginForm-username");
  const passwordInput = container.querySelector("#LoginForm-password");

  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "test-username" } });
    fireEvent.change(passwordInput, { target: { value: "test-password" } });
    fireEvent.submit(form);
  });

  expect(mockLogIn).toHaveBeenCalledWith(
    { username: "test-username", password: "test-password" }
  );
  expect(mockLogIn).toBeCalledTimes(1);

  // shows errors
  const error = container.querySelector(".Error");
  expect(error).toContainHTML("mock error message");
  expect(container).toContainElement(error);

  // type in new inputs and resubmit form
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: "new-username2" } });
    fireEvent.change(passwordInput, { target: { value: "new-password2" } });
    fireEvent.submit(form);
  });

  expect(mockLogIn).toHaveBeenCalledWith(
    { username: "new-username2", password: "new-password2" }
  );
  expect(mockLogIn).toBeCalledTimes(2);

  mockLogIn.mockRestore();
});
