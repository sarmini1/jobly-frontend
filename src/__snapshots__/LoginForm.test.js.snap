// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="LoginForm"
  >
    <div
      class="row"
    >
      <div
        class="col-1 col-sm-2 col-lg-4"
      />
      <div
        class="col-10 col-sm-8 col-lg-4"
      >
        <h3>
          Log in
        </h3>
        <form
          class=""
        >
          <label
            class="form-label"
            for="LoginForm-username"
          >
            Username
          </label>
          <input
            autocomplete="username"
            class="form-control"
            id="LoginForm-username"
            name="username"
            required=""
            type="text"
            value=""
          />
          <label
            class="form-label"
            for="LoginForm-password"
          >
            Password
          </label>
          <input
            autocomplete="current-password"
            class="form-control"
            id="LoginForm-password"
            name="password"
            required=""
            type="password"
            value=""
          />
          <button
            class="btn"
            type="submit"
          >
            Login!
          </button>
        </form>
      </div>
      <div
        class="col-1 col-sm-2 col-lg-4"
      />
    </div>
  </div>
</div>
`;
